### PR TITLE
feat: hashed state provider

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Add `HashedStateProvider` to enable execution by accessing only hashed state.

### CODE REVIEW

**Comment by gakonst** [comment]
Suspect you need an LRU for the keccaks?

File: crates/storage/provider/src/traits/state.rs [hunk 1]
```diff
@@ -12,68 +14,98 @@ pub type StateProviderBox = Box<dyn StateProvider>;
...
+        Ok(self.basic_account(address)?.map(|a| a.nonce))
+    }
 
     /// Get account and storage proofs.
     fn proof(&self, address: Address, keys: &[B256]) -> ProviderResult<AccountProof>;
+}
 
-    /// Get account code by its address.
+/// An abstraction for a type that provides state data for accounts by their hashed address.
+#[auto_impl(&, Arc, Box)]
```
Comment (emhane) [inline crates/storage/provider/src/traits/state.rs:61]: nice, didn't know of this crate. we have it as dep in a lot of crates, maybe we want to move it to the workspace manifest? out of scope nonetheless

**Review by emhane — APPROVED** [review]
lgtm

**Comment by onbjerg** [comment]
Bump @rkrasiuk, what's blocking this from being merged?

**Comment by rkrasiuk** [comment]
@onbjerg some more reviews :) this is not strictly needed atm, but would be nice to have for external users

**Comment by onbjerg** [comment]
gotcha, i'll take a look :)

**Comment by onbjerg** [comment]
Can you expand a bit on the use case? At first I thought this was just a `StateProvider` that hashes the given addresses/slots and then looks them up in the hashed tables, but the hashed provider is a separate construct that reads using pre-hashed keys

**Review by onbjerg — COMMENTED** [review]
Can you expand a bit on the use case? At first I thought this was just a `StateProvider` that hashes the given addresses/slots and then looks them up in the hashed tables, but the hashed provider is a separate construct that reads using pre-hashed keys

**Pull request closed without merging**